hypothesis-python/src/hypothesis/searchstrategy/strings.py

```diff
@@ -63,19 +63,38 @@ def __init__(self,
                 'combination of arguments: ' + ', '.join(
                     '%s=%r' % arg for arg in arguments if arg[1] is not None)
             )
-        self.intervals = IntervalSet(intervals)
-        if whitelist_characters:
-            self.whitelist_characters = set(whitelist_characters)
-        else:
-            self.whitelist_characters = set()
-        self.zero_point = self.intervals.index_above(ord('0'))
+        self.category_intervals = [IntervalSet(i) for i in intervals]
+
+        all_intervals = sorted(sum(intervals, ()), reverse=True)
+        self.all_intervals = IntervalSet(
+            charmap._union_intervals(all_intervals[:1], all_intervals[1:])
+        )
+        self.zero_points = {iset: iset.index_above(ord('0'))
+                            for iset in self.category_intervals}
+        self.zero_points[None] = self.all_intervals.index_above(ord('0'))
 
     def do_draw(self, data):
+        category = choice(data, self.category_intervals)
+        # Almost always - when generating characters - we use the category
+        # chosen above and draw a codepoint from that category.  Then, we
+        # write the integer that would draw this from *all* categories.
+        data.start_example(ONE_UNICODE_CHAR_LABEL)
+        if biased_coin(data, p=0.001):
+            char_idx = category[integer_range(
+                data, 0, len(category) - 1, center=self.zero_points[category]
+            )]
+            data.write(int_to_bytes(self.all_intervals.index(char_idx), 4))
+            data.stop_example()
```

Comment:
These `stop_example` placements aren't quite right.

The idea of "shrink to open" is that the example that is zeroed is you've got something like the following pseudocode:

```python
closed, generated = ... # This should be a single logical example
if closed:
    write(generated)
else:
   generated = draw_normally
return generated
```

The result is that when the original `is_closed, generated` value is zeroed atomically then we have `is_closed` is false and normal drawing kicks in.

So what we would have in the initial generation is `[bytes for boolean true] [bytes for category] [bytes for index into category] [written bytes for code point]`. When zero examples is run we want those first three blocks turned into a single zero block, so we have `[bytes for boolean false] [read bytes for code point]`.

In order to make this work in this case I think you should:

1. Move the category draw inside the if block
2. Close the `stop_example` *before* the write 

Additionally: I'm a little uncomfortable with the use of `p=0.001` right now. I don't think that will work correctly due to some shenanigans with `biased_coin` - right now zeroing it will not work correctly because a false element is two bytes rather than one. I've been meaning to fix that so that zeroing works properly in that case, but that's a separate chunk of work. Suggest replacing it with `data.draw_bits(8) == 0`

Sorry, this is all rather subtle and fiddly. :-( I need to document this trick much better (at all) than it currently is.